agent/consul/gateways/gateway_meta.go

```diff
@@ -17,6 +21,36 @@ type gatewayMeta struct {
 	Gateway *structs.APIGatewayConfigEntry
 }
 
+// getAllGatewayMeta returns a pre-constructed list of all valid gateway and state
+// tuples based on the state coming from the store. Any gateway that does not have
+// a current bound state will be filtered out.
+func getAllGatewayMeta(store *state.Store) ([]*gatewayMeta, error) {
+	_, gateways, err := store.ConfigEntriesByKind(nil, structs.APIGateway, acl.WildcardEnterpriseMeta())
+	if err != nil {
+		return nil, err
+	}
+	_, boundGateways, err := store.ConfigEntriesByKind(nil, structs.BoundAPIGateway, acl.WildcardEnterpriseMeta())
+	if err != nil {
+		return nil, err
+	}
+
+	meta := make([]*gatewayMeta, 0, len(boundGateways))
+	for _, b := range boundGateways {
+		bound := b.(*structs.BoundAPIGatewayConfigEntry)
+		for _, g := range gateways {
+			gateway := g.(*structs.APIGatewayConfigEntry)
+			if bound.References(gateway) {
+				meta = append(meta, &gatewayMeta{
+					BoundGateway: bound,
+					Gateway:      gateway,
+				})
+				break
+			}
+		}
```

Comment:
Might cause more headaches than it's worth, but I think we could pop the gateway at the index after it is matched to make successive checks faster.

Reply:
So, the problem is that the BoundAPIGateway and APIGateway config entries might not align entirely 1-1 or in the same order, for example, maybe a gateway has been deleted, but its bound gateway is still around, or maybe it hasn't been reconciled, or maybe we ended up reconciling gateways in the opposite order that they were created due to a transient reconciliation failure. Basically, this code is trying to handle any number of gateways in any order that may or may not have corresponding bound gateways, so I don't think we can just use the same index.

Reply:
Ah, nvm, I got what you're saying, reslice gateways so we don't have to iterate over the whole length of the array. I'd be alright with that, but I think it's probably more trouble than it's worth as an initial optimization.

Reply:
I agree. We can come back to it if this becomes a pain point, but it's not likely.